# Analysing the 'BlogFeedback Data Set' from the UC Irvine Machine Learning repository

This notebook is used to analyze the 'BlogFeedback Data Set' from the UC Irvine Machine Learning repository. The data set is available [here](https://archive.ics.uci.edu/ml/datasets/BlogFeedback). **The objective of the notebook is to create a model to predict the number of comments in a blog post in the upcoming 24 hours**.

This data originates from blog posts. The raw HTML-documents of the blog posts were crawled and processed. In the train data, the basetimes were in the years 2010 and 2011. In the test data the basetimes were in February and March 2012.

**The data set has 280 attributes. Therefore, in this notebooks we test different techniques to deal with this large number of attributes**. First, we analyze the whole data set without any kind of adjustment. This will be our reference model. Then, we test some feature selection methods to identify the most relevant attributes to predict the target value. Finally, we test the Principal Component Analysis (PCA) dimensionality reduction method.

The notebook is organized as follows:

1. Data exploration
2. Train ML model
3. Evaluate the ML model
4. Conclusion

----------

## 1. Data exploration

In this section, we explore the characteristics of the data set, including its dimensions and characteristics of its variables.

The data set contains 281 columns and 52397 rows.

The attributes of the data set are the following:

Column:
- 1...50: Average, standard deviation, min, max and median of the Attributes 51...60 for the source of the current blog post. With source we mean the blog on which the post appeared. For example, myblog.blog.org would be the source of the post myblog.blog.org/post_2010_09_10
- 51: Total number of comments before basetime
- 52: Number of comments in the last 24 hours before the basetime
- 53: Let T1 denote the datetime 48 hours before basetime. Let T2 denote the datetime 24 hours before basetime. This attribute is the number of comments in the time period between T1 and T2
- 54: Number of comments in the first 24 hours after the publication of the blog post, but before basetime
- 55: The difference of Attribute 52 and Attribute 53
- 56...60: The same features as the attributes 51...55, but features 56...60 refer to the number of links (trackbacks), while features 51...55 refer to the number of comments.
- 61: The length of time between the publication of the blog post and basetime
- 62: The length of the blog post
- 63...262: The 200 bag of words features for 200 frequent words of the text of the blog post
- 263...269: binary indicator features (0 or 1) for the weekday (Monday...Sunday) of the basetime
- 270...276: binary indicator features (0 or 1) for the weekday (Monday...Sunday) of the date of publication of the blog post
- 277: Number of parent pages: we consider a blog post P as a parent of blog post B, if B is a reply (trackback) to blog post P.
- 278...280: Minimum, maximum, average number of comments that the parents received
- 281: The target: the number of comments in the next 24 hours (relative to basetime)

In [2]:
import pandas as pd
import numpy as np
#!pip install -U scikit-learn

----------

### Getting the data

In [3]:
attributes = [*range(1, 282, 1)]

df_data = pd.read_csv('/Users/leuzinger/Dropbox/Data Science/Awari/Regressions/BlogFeedback/blogData_train.csv',names=attributes)
df_data.reset_index(inplace=False)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52397 entries, 0 to 52396
Columns: 281 entries, 1 to 281
dtypes: float64(281)
memory usage: 112.3 MB


In [4]:
att=[]
for i in ['total','last24h','24-48h','first24h','difference',
           'total_tr','last24h_tr','24-48h_tr','first24h_tr','difference_tr']:
    att1 = 'blog_avg_' + str(i)
    att2 = 'blog_std_' + str(i)
    att3 = 'blog_min_' + str(i)
    att4 = 'blog_max_' + str(i)
    att5 = 'blog_median_' + str(i)
    att.extend([att1,att2,att3,att4,att5])

att51_62 = ['total','last24h','24-48h','first24h','difference',
           'total_tr','last24h_tr','24-48h_tr','first24h_tr','difference_tr',
           'time_first_post','lenght_post']
att.extend(att51_62)

for i in range(63,263):
    att_word = 'word' + str(i-62)
    att.extend([att_word])

att263_281 = ['Mon_bl','Tue_bl','Wed_bl','Thu_bl','Fri_bl','Sat_bl','Sun_bl',
             'Mon_post','Tue_post','Wed_post','Thu_post','Fri_post','Sat_post','Sun_post',
             'parent_pages','min_parent','max_parent','avg_parent','target']
att.extend(att263_281)

In [5]:
df_data.set_axis(att,axis=1,inplace=True)
df_data.head()

blog_avg_total  blog_std_total  blog_min_total  blog_max_total  \
0        40.30467       53.845657             0.0           401.0   
1        40.30467       53.845657             0.0           401.0   
2        40.30467       53.845657             0.0           401.0   
3        40.30467       53.845657             0.0           401.0   
4        40.30467       53.845657             0.0           401.0   

   blog_median_total  blog_avg_last24h  blog_std_last24h  blog_min_last24h  \
0               15.0          15.52416          32.44188               0.0   
1               15.0          15.52416          32.44188               0.0   
2               15.0          15.52416          32.44188               0.0   
3               15.0          15.52416          32.44188               0.0   
4               15.0          15.52416          32.44188               0.0   

   blog_max_last24h  blog_median_last24h  blog_avg_24-48h  blog_std_24-48h  \
0             377.0                  3.0        14.044226        32.615417   
1             377.0                  3.0        14.044226        32.615417   
2             377.0                  3.0        14.044226        32.615417   
3             377.0                  3.0        14.044226        32.615417   
4             377.0                  3.0        14.044226        32.615417   

   blog_min_24-48h  blog_max_24-48h  blog_median_24-48h  blog_avg_first24h  \
0              0.0            377.0                 2.0          34.567566   
1              0.0            377.0                 2.0          34.567566   
2              0.0            377.0                 2.0          34.567566   
3              0.0            377.0                 2.0          34.567566   
4              0.0            377.0                 2.0          34.567566   

   blog_std_first24h  blog_min_first24h  blog_max_first24h  \
0          48.475178                0.0              378.0   
1          48.475178                0.0              378.0   
2          48.475178                0.0              378.0   
3          48.475178                0.0              378.0   
4          48.475178                0.0              378.0   

   blog_median_first24h  blog_avg_difference  blog_std_difference  \
0                  12.0             1.479934             46.18691   
1                  12.0             1.479934             46.18691   
2                  12.0             1.479934             46.18691   
3                  12.0             1.479934             46.18691   
4                  12.0             1.479934             46.18691   

   blog_min_difference  blog_max_difference  blog_median_difference  \
0               -356.0                377.0                     0.0   
1               -356.0                377.0                     0.0   
2               -356.0                377.0                     0.0   
3               -356.0                377.0                     0.0   
4               -356.0                377.0                     0.0   

   blog_avg_total_tr  blog_std_total_tr  blog_min_total_tr  blog_max_total_tr  \
0           1.076167           1.795416                0.0               11.0   
1           1.076167           1.795416                0.0               11.0   
2           1.076167           1.795416                0.0               11.0   
3           1.076167           1.795416                0.0               11.0   
4           1.076167           1.795416                0.0               11.0   

   blog_median_total_tr  blog_avg_last24h_tr  blog_std_last24h_tr  \
0                   0.0             0.400491             1.078097   
1                   0.0             0.400491             1.078097   
2                   0.0             0.400491             1.078097   
3                   0.0             0.400491             1.078097   
4                   0.0             0.400491             1.078097   

   blog_min_last24h_tr  blog_max_last24h_tr  blog_median_last24h

----------

### Data correlation

We start our analysis looking to which attributes have the higher correlation with the price. First, we create a correlation matrix. 

We can see that the varibales that have the stronger postive correlations with the target value are the blog_median_last24h and blog_avg_difference. Besides, we see that the total blog publications, publications in the last 24h and publications between 24h-48h have a strong correlation with each other.

In [6]:
corr_matrix = df_data.corr()
corr_matrix.loc['target'].sort_values(ascending=False)

target                       1.000000
blog_median_last24h          0.506540
blog_avg_difference          0.503375
blog_avg_last24h             0.497631
blog_median_total            0.491707
blog_avg_24-48h              0.490111
blog_median_24-48h           0.489674
blog_median_first24h         0.486316
blog_avg_total               0.485464
last24h                      0.472061
blog_avg_first24h            0.471999
blog_median_last24h_tr       0.461627
blog_std_difference          0.440003
blog_std_24-48h              0.439152
blog_std_last24h             0.433578
blog_std_total               0.424616
blog_std_first24h            0.384654
blog_max_total               0.356604
blog_median_total_tr         0.338961
blog_avg_24-48h_tr           0.337775
blog_avg_last24h_tr          0.335829
blog_avg_first24h_tr         0.329670
blog_avg_total_tr            0.328525
blog_median_first24h_tr      0.323661
blog_max_24-48h              0.322775
blog_max_last24h             0.322106
blog_max_dif

----------

### Creating the Train and Test sets

Creating a test set at the beginning of the project avoid *data snooping* bias, i.e., "when you estimate the generalization error using the test set, your estimate will be too optimistic, and you will launch a system that will not perform as well as expected" (GÉRON, 2019).

In this data set, the test set has already been divided. Therefore, we do not need to create a test set, just separete the target value from the other attributes to create our training set.

In [13]:
blog_X_train = df_data.drop('target',axis=1).copy()
blog_y_train = df_data['target'].copy()

In [14]:
df_blog = df_data.iloc[:, :50].copy()
df_comments = df_data.iloc[:, 50:62].copy()
df_words = df_data.iloc[:, 62:262].copy()
df_weekday = df_data.iloc[:, 262:276].copy()
df_parent = df_data.iloc[:, 276:280].copy()

----------

### Preparing the data for ML algorithms

Before creating the ML models, we need to prepare the data so that the ML algorithms will work properly.

First, we need to clean missing values from the dataset. Second, we need to put all the attributes in the same scale because "Machine Learning algorithms don’t perform well when the input numerical attributes have very different scales" [(GÉRON, 2019)](https://www.amazon.com.br/Hands-Machine-Learning-Scikit-Learn-TensorFlow/dp/1492032646).

We verify that there is no missing values in our data set. So, we just prepare a pipeline to do the scaling when necessary.

In [15]:
df_blog.isnull().values.any()

False

In [197]:
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

def estimator_transf(estimator):
    #imputer = SimpleImputer(strategy='median')
    pipeline = Pipeline(steps=[('m', estimator)])
    return pipeline

def estimator_scaler(estimator):
    #imputer = SimpleImputer(strategy='median')
    pipeline = Pipeline(steps=[('scaler',StandardScaler()),('model', estimator)])
    return pipeline 

----------

## 2. Train ML model

After preparing the data set, we are ready to select and train our ML model.

We start with a Linear Regression (LR) model. "A regression model, such as linear regression, models an output value based on a linear combination of input values" [(BROWNLEE, 2020)](https://machinelearningmastery.com/introduction-to-time-series-forecasting-with-python/).

Then, we try some regularized linear models. This kind of model constrain the weights of the model, avoiding overfitting (GÉRON, 2019). We try three regularized linear models [(BROWNLEE, 2016)](https://machinelearningmastery.com/machine-learning-with-python/):

1. Ridge regression. This model model uses the L2 regularization. It adds “squared magnitude” of coefficient as a penalty term to the loss function [(NAGPAL, 2017)](https://towardsdatascience.com/l1-and-l2-regularization-methods-ce25e7fc831c).
2. Lasso regression. This model model uses the L1 regularization. It adds “absolute value of magnitude” of coefficient as penalty term to the loss function (NAGPAL, 2017).
3. Elastic Net. This model combines the Ridge and the Lasso models. "It seeks to minimize the complexity of the regression model (magnitude and number of regression coefficients) by penalizing the model using both the L2-norm (sum squared coefficient values) and the L1-norm (sum absolute coefficient values)" (BROWNLEE, 2016).

Finally, we also try some nonlinear algorithms:

1. Classification and Regression Trees (CART). It uses "the train- ing data to select the best points to split the data in order to minimize a cost metric" (BROWNLEE, 2016).
2. k-Nearest Neighbors (KNN). This model "locates the k most similar instances in the training dataset for a new data instance" (BROWNLEE, 2016).

The models are evaluated using the mean absolute error (MAE), root square mean error (RMSE), and R². RMSE punish larger errors more than smaller errors, inflating or magnifying the mean error score. This is due to the square of the error value. MAE does not give more or less weight to different types of errors and instead the scores increase linearly with increases in error. MAE is the simplest evaluation metric and most easily interpreted. R² tells you how much variance your model accounts for. In the case of the MAE and RMSE, the lower the better. But for R², the close the value is to 1, the better ([HALE, 2020](https://towardsdatascience.com/which-evaluation-metric-should-you-use-in-machine-learning-regression-problems-20cdaef258e); [BROWNLEE, 2021](https://machinelearningmastery.com/regression-metrics-for-machine-learning/)).

Besides, "the key to a fair comparison of machine learning algorithms is ensuring that each algorithm is evaluated in the same way on the same data. You can achieve this by forcing each algorithm to be evaluated on a consistent test harness" (BROWNLEE, 2016). In this project, we do this by using the same split in the cross validation. We use the KFold function from the sklearn library with a random value rs as the random_state parameter. Although the rs value change everytime the notebook is run, once it is set, the same rs value is used in all the models. This guarantees that all the models are evaluated on the same data.

The result of the tests of the models with the training data shows that **the KNN is the best model**. It has the lowest MAE and RMSE, and the highest R².

However, differing scales of the raw data could be negatively impacting the performance of some of the models. Therefore, we test the models again, but this time we standardize the data set.

We can see that the performance of most models improved with standardization. However, the performance of the KNN degraded with the standardized data. Even so, KNN was still the best method.

**Therefore, for this initial test, we verify that KNN without standardization is the best model for our data**.

However, **using the data set with all the 280 attributes requires a lot of computing time**. So, let's try some featuring selection methods to see if we can reduce the number of attributes to be used in our models.

In [17]:
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_validate
from sklearn.model_selection import KFold

def estimator_cross_val (model,estimator,pipe,matriz,rs,X,y):
    pipe_ = pipe(estimator)
    scoring = ['neg_mean_absolute_error', 'neg_root_mean_squared_error','r2']
    kfold = KFold(n_splits=5, random_state=rs,shuffle=True)
    scores = cross_validate(pipe_,X,y,cv=kfold,scoring=scoring)
    
    mae_scores = -scores.get('test_neg_mean_absolute_error')
    mae_mean = mae_scores.mean()
    mae_std = mae_scores.std()
    
    rmse_scores = -scores.get('test_neg_root_mean_squared_error')
    rmse_mean = rmse_scores.mean()
    rmse_std = rmse_scores.std()
    
    r2_scores = scores.get('test_r2')
    r2_mean = r2_scores.mean()
    r2_std = r2_scores.std()
    
    results_ = [model,mae_mean,mae_std,rmse_mean,rmse_std,r2_mean,r2_std]
    results_ = pd.Series(results_, index = matriz.columns)
    results = matriz.append(results_,ignore_index=True)
    return results

In [18]:
from random import randrange
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.linear_model import ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
import warnings

warnings.filterwarnings("ignore")

rs = randrange(10000)
matriz = pd.DataFrame(columns=['model','MAE_mean','MAE_std','RMSE_mean','RMSE_std','R2_mean','R2_std'])

matriz = estimator_cross_val('Linear Regression',LinearRegression(),estimator_transf,matriz,rs,blog_X_train,blog_y_train)
matriz = estimator_cross_val('Ridge Regression',Ridge(),estimator_transf,matriz,rs,blog_X_train,blog_y_train)
matriz = estimator_cross_val('Lasso',Lasso(),estimator_transf,matriz,rs,blog_X_train,blog_y_train)
matriz = estimator_cross_val('Elastic Net',ElasticNet(),estimator_transf,matriz,rs,blog_X_train,blog_y_train)
matriz = estimator_cross_val('KNN',KNeighborsRegressor(),estimator_transf,matriz,rs,blog_X_train,blog_y_train)
matriz = estimator_cross_val('CART',DecisionTreeRegressor(),estimator_transf,matriz,rs,blog_X_train,blog_y_train)
matriz

model  MAE_mean   MAE_std  RMSE_mean  RMSE_std   R2_mean  \
0  Linear Regression  9.544545  0.103498  30.381871  2.102113  0.348687   
1   Ridge Regression  9.533756  0.101103  30.373981  2.102229  0.349013   
2              Lasso  9.131463  0.096692  30.283813  2.139727  0.352875   
3        Elastic Net  9.143876  0.094120  30.297370  2.129937  0.352289   
4                KNN  6.390140  0.188943  28.577260  1.109094  0.421754   
5               CART  6.417226  0.272852  34.032947  3.212304  0.175991   

     R2_std  
0  0.046967  
1  0.047121  
2  0.048578  
3  0.048275  
4  0.043830  
5  0.143207

In [19]:
matriz2 = pd.DataFrame(columns=['model','MAE_mean','MAE_std','RMSE_mean','RMSE_std','R2_mean','R2_std'])

matriz2 = estimator_cross_val('Linear Regression',LinearRegression(),estimator_scaler,matriz2,rs,blog_X_train,blog_y_train)
matriz2 = estimator_cross_val('Ridge Regression',Ridge(),estimator_scaler,matriz2,rs,blog_X_train,blog_y_train)
matriz2 = estimator_cross_val('Lasso',Lasso(),estimator_scaler,matriz2,rs,blog_X_train,blog_y_train)
matriz2 = estimator_cross_val('Elastic Net',ElasticNet(),estimator_scaler,matriz2,rs,blog_X_train,blog_y_train)
matriz2 = estimator_cross_val('KNN',KNeighborsRegressor(),estimator_scaler,matriz2,rs,blog_X_train,blog_y_train)
matriz2 = estimator_cross_val('CART',DecisionTreeRegressor(),estimator_scaler,matriz2,rs,blog_X_train,blog_y_train)
matriz2

model  MAE_mean   MAE_std  RMSE_mean  RMSE_std   R2_mean  \
0  Linear Regression  9.547968  0.103276  30.381809  2.099744  0.348691   
1   Ridge Regression  9.539233  0.102446  30.375292  2.099088  0.348960   
2              Lasso  8.441711  0.109987  30.462235  2.257990  0.345548   
3        Elastic Net  8.448227  0.094508  30.637003  2.219413  0.338140   
4                KNN  6.784293  0.155734  29.294693  2.056182  0.394841   
5               CART  6.366500  0.264876  34.032960  2.777628  0.177901   

     R2_std  
0  0.046852  
1  0.046955  
2  0.049525  
3  0.046236  
4  0.039818  
5  0.118054

-----

### Feature selection

"Feature selection is the process of reducing the number of input variables when developing a predictive model. It is desirable to reduce the number of input variables to both reduce the computational cost of modeling and, in many cases, to improve the performance of the model" (BROWNLEE, 2021).

There are two main techniques of feature selection: supervised and unsupervised. Supervised methods use the target variable, while unsupervised methods do not (BROWNLEE, 2021).

Besides, the supervised techniques can be divided in (BROWNLEE, 2021):

1. Intrinsic: Algorithms that perform automatic feature selection during training.
2. Wrapper: Search subsets of features that perform according to a predictive model.
3. Filter: Select subsets of features based on their relationship with the target.

### Mutual Information Statistics

Some of the methods of feature selection are more appropriated for numerical variables and others for categorical ones. One popular feature selection techniques used for both numerical variables and categorical variable is Mutual Information Statistics (BROWNLEE, 2021). 

"Mutual information from the field of information theory is the application of information gain (typically used in the construction of decision trees) to feature selection. Mutual information is calculated between two variables and measures the reduction in uncertainty for one variable given a known value of the other variable"  (BROWNLEE, 2021).

We find that many attributes have negligible information value. 181 features have a contribution score over 0.0001, 144 over 0.001, 65 over 0.01, and only 33 over 0.1. **These numbers can vary depending on the training set**. Therefore, we will test the 30, 70, 150, and 190 best features and compare it with the results obtained using all features. 

**We see that the performance using the 70, 150, and 190 best features are almost the same of using all features. Using the 30 beast features is just slighlty worst than using all features**. Moreover, in all cases the KNN model have the best performance.

We could do a grid search to "systematically test a range of different numbers of selected features and discover which results in the best performing model" (BROWNLEE, 2021). **However, a grid search to determine the optimum number of features would require a lot of computing time and the benefit would not be significant in our evaluation**.

In [20]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import mutual_info_classif

# feature selection
def select_features(X_train, y_train,k_):
    # configure to select all features
    fs = SelectKBest(score_func=mutual_info_classif, k=k_) 
    # learn relationship from training data
    fs.fit(X_train, y_train)
    # transform train input data
    X_train_fs = fs.transform(X_train)
    return X_train_fs, fs

In [22]:
# feature selection
blog_X_train_mi, mi = select_features(blog_X_train, blog_y_train,'all')

# what are scores for the features
MI = pd.DataFrame(mi.scores_, columns = ['Score'])
MI.sort_values(by=['Score'],ascending=False)

Score
51   0.275880
54   0.237840
21   0.209719
11   0.208136
0    0.207897
6    0.207496
1    0.206831
10   0.203635
20   0.202529
5    0.202373
16   0.200722
15   0.194932
23   0.191542
8    0.190217
13   0.188195
3    0.186027
22   0.184062
18   0.183795
4    0.178890
50   0.174909
19   0.173680
53   0.170095
9    0.168995
14   0.135486
31   0.119140
26   0.118753
25   0.115724
46   0.114905
36   0.114312
35   0.112838
41   0.108989
30   0.107790
40   0.104771
45   0.076726
61   0.073060
245  0.068306
142  0.065440
100  0.064659
48   0.064550
60   0.063339
43   0.058855
28   0.058288
24   0.058171
52   0.056459
33   0.054157
38   0.050571
47   0.050074
212  0.039152
231  0.030719
66   0.030007
157  0.029748
29   0.028234
56   0.025834
225  0.024111
44   0.022696
34   0.020177
138  0.018655
68   0.018598
59   0.017393
209  0.017102
71   0.015419
55   0.015041
78   0.012075
58   0.012028
113  0.010003
247  0.009429
232  0.008656
2    0.008284
150  0.008241
127  0.008227
103  0.008166
63   0.007725
121  0.007161
67   0.007090
193  0.006739
201  0.006684
119  0.006538
190  0.005337
95   0.005262
87   0.005123
217  0.005095
246  0.005080
182  0.004222
233  0.004177
82   0.004174
76   0.004020
199  0.003978
203  0.003943
265  0.003709
102  0.003610
81   0.003481
172  0.003452
98   0.003340
183  0.003334
222  0.003280
187  0.003264
42   0.003231
241  0.003169
93   0.003117
270  0.003099
131  0.003050
122  0.003002
254  0.002906
269  0.002892
17   0.002869
206  0.002739
80   0.002711
144  0.002444
118  0.002391
271  0.002224
267  0.002220
74   0.002188
218  0.002179
192  0.002174
132  0.002167
137  0.002078
260  0.002067
116  0.002047
133  0.001990
250  0.001936
166  0.001913
79   0.001901
149  0.001864
197  0.001798
97   0.001750
196  0.001731
162  0.001703
92   0.001695
174  0.001625
85   0.001555
239  0.001504
189  0.001477
177  0.001435
12   0.001363
169  0.001286
49   0.001281
91   0.001257
264  0.001187
276  0.001180
39   0.001178
159  0.001165
184  0.001161
130  0.001088
191  0.001015
213  0.000970
173  0.000967
272  0.000894
186  0.000875
120  0.000872
216  0.000865
236  0.000846
238  0.000800
185  0.000799
202  0.000788
141  0.000782
256  0.000717
96   0.000699
200  0.000693
198  0.000623
32   0.000616
107  0.000606
145  0.000592
140  0.000591
112  0.000542
208  0.000538
273  0.000501
240  0.000478
143  0.000476
126  0.000475
129  0.000462
210  0.000364
237  0.000287
179  0.000266
139  0.000240
278  0.000234
153  0.000229
72   0.000219
228  0.000212
230  0.000196
171  0.000155
252  0.000110
257  0.000090
147  0.000044
234  0.000024
215  0.000021
83   0.000000
84   0.000000
244  0.000000
77   0.000000
163  0.000000
243  0.000000
242  0.000000
249  0.000000
86   0.000000
88   0.000000
89   0.000000
90   0.000000
235  0.000000
94   0.000000
99   0.000000
101  0.000000
248  0.000000
73   0.000000
75   0.000000
251  0.000000
277  0.000000
7    0.000000
275  0.000000
274  0.000000
27   0.000000
37   0.000000
57   0.000000
62   0.000000
268  0.000000
64   0.000000
266  0.000000
65   0.000000
69   0.000000
263  0.000000
262  0.000000
261  0.000000
70   0.000000
259  0.000000
258  0.000000
255  0.000000
253  0.000000
229  0.000000
221  0.000000
227  0.000000
226  0.000000
134  0.000000
135  0.000000
136  0.000000
188  0.000000
146  0.000000
148  0.000000
151  0.000000
152  0.000000
181  0.000000
180  0.000000
178  0.000000
154  0.000000
176  0.000000
175  0.000000
155  0.000000
156  0.000000
158  0.000000
170  0.000000
160  0.000000
168  0.000000
167  0.000000
161  0.000000
165  0.000000
194  0.000000
195  0.000000
128  0.000000
214  0.000000
104  0.000000
224  0.000000
223  0.000000
105  0.000000
164  0.000000
220  0.000000
219  0.000000
106  0.000000
108  0.000000
109  0.000000
110  0.000000
125  0.000000
111  0.000000
211  0.000000
207  0.000000
114  0.000000
205  0.000000
204  0.000000
115  0.000000
117  0.000000
123  0.000000
124  0.000000
279  0.000000

In [23]:
print(MI[MI > 0.0001].count())
print(MI[MI > 0.001].count())
print(MI[MI > 0.01].count())
print(MI[MI > 0.1].count())

Score    181
dtype: int64
Score    144
dtype: int64
Score    65
dtype: int64
Score    33
dtype: int64


In [113]:
def estimator_cross_val_fea (k,model,estimator,pipe,matriz,rs,X,y):
    pipe_ = pipe(estimator)
    scoring = ['neg_mean_absolute_error', 'neg_root_mean_squared_error','r2']
    kfold = KFold(n_splits=5, random_state=rs,shuffle=True)
    scores = cross_validate(pipe_,X,y,cv=kfold,scoring=scoring)
    
    mae_scores = -scores.get('test_neg_mean_absolute_error')
    mae_mean = mae_scores.mean()
    mae_std = mae_scores.std()
    
    rmse_scores = -scores.get('test_neg_root_mean_squared_error')
    rmse_mean = rmse_scores.mean()
    rmse_std = rmse_scores.std()
    
    r2_scores = scores.get('test_r2')
    r2_mean = r2_scores.mean()
    r2_std = r2_scores.std()
    
    results_ = [k,model,mae_mean,mae_std,rmse_mean,rmse_std,r2_mean,r2_std]
    results_ = pd.Series(results_, index = matriz.columns)
    results = matriz.append(results_,ignore_index=True)
    return results

matriz_mi = pd.DataFrame(columns=['features','model','MAE_mean','MAE_std','RMSE_mean','RMSE_std','R2_mean','R2_std'])

for k in [30,70,150,190]:

    best_features_mi = MI.transpose()
    best_features_mi.columns = blog_X_train.columns
    best_features_mi.sort_values('Score',axis=1,ascending=False,inplace=True)
    best_features_mi.drop(best_features_mi.iloc[:,k:],axis=1,inplace=True)
    blog_X_train_mi = blog_X_train[best_features_mi.columns]
  
    matriz_mi = estimator_cross_val_fea(k,'Linear Regression',LinearRegression(),     estimator_transf,matriz_mi,rs,blog_X_train_mi,blog_y_train)
    matriz_mi = estimator_cross_val_fea(k,'Ridge Regression', Ridge(),                estimator_transf,matriz_mi,rs,blog_X_train_mi,blog_y_train)
    matriz_mi = estimator_cross_val_fea(k,'Lasso',            Lasso(),                estimator_transf,matriz_mi,rs,blog_X_train_mi,blog_y_train)
    matriz_mi = estimator_cross_val_fea(k,'Elastic Net',      ElasticNet(),           estimator_transf,matriz_mi,rs,blog_X_train_mi,blog_y_train)
    matriz_mi = estimator_cross_val_fea(k,'KNN',              KNeighborsRegressor(),  estimator_transf,matriz_mi,rs,blog_X_train_mi,blog_y_train)
    matriz_mi = estimator_cross_val_fea(k,'CART',             DecisionTreeRegressor(),estimator_transf,matriz_mi,rs,blog_X_train_mi,blog_y_train)

matriz_mi

features              model  MAE_mean   MAE_std  RMSE_mean  RMSE_std  \
0        30  Linear Regression  8.102044  0.070500  30.460122  2.164337   
1        30   Ridge Regression  8.098280  0.073934  30.453557  2.162956   
2        30              Lasso  8.102536  0.087173  30.402198  2.179639   
3        30        Elastic Net  8.111061  0.087429  30.411269  2.179164   
4        30                KNN  6.416703  0.230324  28.668569  0.863457   
5        30               CART  7.111597  0.445084  35.918582  3.811208   
6        70  Linear Regression  9.261467  0.091643  30.344233  2.128451   
7        70   Ridge Regression  9.257935  0.088683  30.340661  2.127563   
8        70              Lasso  9.127167  0.096721  30.287729  2.135822   
9        70        Elastic Net  9.148627  0.091839  30.304993  2.126502   
10       70                KNN  6.405011  0.194332  28.583471  1.106401   
11       70               CART  6.451796  0.220880  33.634105  2.733972   
12      150  Linear Regression  9.382456  0.094181  30.351116  2.130122   
13      150   Ridge Regression  9.378137  0.092659  30.347295  2.128722   
14      150              Lasso  9.127167  0.096721  30.287729  2.135822   
15      150        Elastic Net  9.148563  0.091806  30.305229  2.126215   
16      150                KNN  6.395976  0.193652  28.579396  1.103361   
17      150               CART  6.449221  0.291889  34.058084  3.144257   
18      190  Linear Regression  9.438096  0.095297  30.359550  2.123977   
19      190   Ridge Regression  9.431912  0.093154  30.355105  2.122739   
20      190              Lasso  9.127223  0.096909  30.287813  2.135734   
21      190        Elastic Net  9.148673  0.091982  30.305372  2.126077   
22      190                KNN  6.385101  0.183604  28.571782  1.106768   
23      190               CART  6.571310  0.220003  35.346713  2.761287   

     R2_mean    R2_std  
0   0.345266  0.050290  
1   0.345538  0.050354  
2   0.347780  0.050551  
3   0.347399  0.050441  
4   0.417630  0.044628  
5   0.082226  0.171896  
6   0.350294  0.048114  
7   0.350440  0.048170  
8   0.352686  0.048690  
9   0.351940  0.048425  
10  0.421494  0.043878  
11  0.195243  0.126025  
12  0.349996  0.048221  
13  0.350154  0.048236  
14  0.352686  0.048690  
15  0.351930  0.048413  
16  0.421668  0.043698  
17  0.175378  0.137336  
18  0.349629  0.048041  
19  0.349813  0.048077  
20  0.352683  0.048687  
21  0.351924  0.048409  
22  0.421986  0.043620  
23  0.114920  0.111548

### Wrapper feature selection method

One way to handle data sets that combines numerical and categorical variables is to use a wrapper method. Some ofent used wrapper methods are Tree-Searching Methods, Stochastic Global Search, Step-Wise Models, and Recursive Feature Elimination (BROWNLEE, 2021).

We use the **Recursive Feature Elimination (RFE) method**. This method searches "for a subset of features by starting with all features in the training dataset and successfully removing features until the desired number remains. This is achieved by fitting the given machine learning algorithm used in the core of the model, ranking features by importance, discarding the least important features, and re-fitting the model" (BROWNLEE, 2021).

We use the RFE to reduce the attributes of the data set. We use the same number of features as in the Mutual information selection method. Thus, we select the 190, 150, 70, and 30 most relevant features and evaluate the models again. 

**We can see that the best model is the KNN with 30 features**. Moreover, the models performed better with the features selected using the RFE than with the ones selected using the mutual information. This model even performed better than the one using all features.

In [166]:
from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeRegressor

matriz_rfe = pd.DataFrame(columns=['features','model','MAE_mean','MAE_std','RMSE_mean','RMSE_std','R2_mean','R2_std'])
rfe_features = pd.DataFrame()

for k in [30,70,150,190]:
    rfe = RFE(estimator=DecisionTreeRegressor(), n_features_to_select=k)
    rfe.fit(blog_X_train,blog_y_train)
    RF = pd.DataFrame(rfe.support_, columns = ['{} Features'.format(k)])
    rfe_features['{} Features'.format(k)] = RF['{} Features'.format(k)]
    
    best_features_rfe = RF.transpose()
    best_features_rfe.columns = blog_X_train.columns
    best_features_rfe.sort_values('{} Features'.format(k),axis=1,ascending=False,inplace=True)
    best_features_rfe.drop(best_features_rfe.iloc[:,k:],axis=1,inplace=True)
    blog_X_train_rfe = blog_X_train[best_features_rfe.columns]
    blog_X_train_rfe.head()
  
    matriz_rfe = estimator_cross_val_fea(k,'Linear Regression',LinearRegression(),     estimator_transf,matriz_rfe,rs,blog_X_train_rfe,blog_y_train)
    matriz_rfe = estimator_cross_val_fea(k,'Ridge Regression', Ridge(),                estimator_transf,matriz_rfe,rs,blog_X_train_rfe,blog_y_train)
    matriz_rfe = estimator_cross_val_fea(k,'Lasso',            Lasso(),                estimator_transf,matriz_rfe,rs,blog_X_train_rfe,blog_y_train)
    matriz_rfe = estimator_cross_val_fea(k,'Elastic Net',      ElasticNet(),           estimator_transf,matriz_rfe,rs,blog_X_train_rfe,blog_y_train)
    matriz_rfe = estimator_cross_val_fea(k,'KNN',              KNeighborsRegressor(),  estimator_transf,matriz_rfe,rs,blog_X_train_rfe,blog_y_train)
    matriz_rfe = estimator_cross_val_fea(k,'CART',             DecisionTreeRegressor(),estimator_transf,matriz_rfe,rs,blog_X_train_rfe,blog_y_train)

matriz_rfe

features              model  MAE_mean   MAE_std  RMSE_mean  RMSE_std  \
0        30  Linear Regression  9.180081  0.104516  30.279467  2.203382   
1        30   Ridge Regression  9.180040  0.104518  30.279461  2.203382   
2        30              Lasso  9.092673  0.101425  30.280741  2.212499   
3        30        Elastic Net  9.122076  0.101398  30.286636  2.211027   
4        30                KNN  6.374234  0.180121  28.476245  1.238052   
5        30               CART  6.232891  0.443807  33.024775  3.559776   
6        70  Linear Regression  9.306786  0.082999  30.294309  2.150059   
7        70   Ridge Regression  9.306366  0.083161  30.293986  2.150332   
8        70              Lasso  9.111878  0.111197  30.249388  2.163756   
9        70        Elastic Net  9.128702  0.111946  30.257852  2.165584   
10       70                KNN  6.365024  0.182862  28.642718  1.190666   
11       70               CART  6.266452  0.332824  33.588791  2.617158   
12      150  Linear Regression  9.447957  0.068327  30.366726  2.122330   
13      150   Ridge Regression  9.439724  0.064408  30.363983  2.122545   
14      150              Lasso  9.126103  0.099535  30.287416  2.131768   
15      150        Elastic Net  9.145160  0.097843  30.300500  2.125547   
16      150                KNN  6.380200  0.191840  28.547150  1.068192   
17      150               CART  6.346948  0.292606  33.734198  2.954824   
18      190  Linear Regression  9.509926  0.091557  30.371558  2.114462   
19      190   Ridge Regression  9.503399  0.089295  30.368305  2.114046   
20      190              Lasso  9.130458  0.097242  30.284351  2.139953   
21      190        Elastic Net  9.144115  0.094547  30.297607  2.129799   
22      190                KNN  6.390132  0.188479  28.577314  1.109036   
23      190               CART  6.399573  0.285867  34.426479  3.186690   

     R2_mean    R2_std  
0   0.353323  0.047883  
1   0.353324  0.047883  
2   0.353230  0.048760  
3   0.352968  0.048825  
4   0.426645  0.034890  
5   0.225693  0.140368  
6   0.352496  0.048134  
7   0.352510  0.048142  
8   0.354344  0.049547  
9   0.354014  0.049239  
10  0.419439  0.041009  
11  0.198748  0.115628  
12  0.349323  0.047963  
13  0.349429  0.048119  
14  0.352719  0.048275  
15  0.352153  0.048116  
16  0.422913  0.043830  
17  0.192542  0.121404  
18  0.349141  0.047317  
19  0.349271  0.047422  
20  0.352853  0.048573  
21  0.352279  0.048264  
22  0.421752  0.043827  
23  0.158403  0.135199

In [167]:
rfe_features

30 Features  70 Features  150 Features  190 Features
0          False        False          True          True
1          False         True          True          True
2          False        False          True          True
3           True         True          True          True
4           True         True          True          True
5          False         True          True          True
6          False        False          True          True
7          False        False         False         False
8          False        False          True          True
9          False         True          True          True
10         False        False          True          True
11         False        False         False          True
12         False        False         False         False
13         False        False          True          True
14         False        False          True          True
15          True         True          True          True
16         False        False          True          True
17         False        False         False          True
18         False        False          True          True
19         False        False          True          True
20          True         True          True          True
21          True         True          True          True
22         False         True          True          True
23         False        False          True          True
24         False         True          True          True
25         False        False          True          True
26         False        False          True          True
27         False        False         False         False
28         False        False         False          True
29         False        False         False          True
30         False        False          True          True
31         False        False          True          True
32         False        False         False         False
33         False        False         False          True
34         False        False         False         False
35         False         True         False          True
36         False        False         False          True
37         False        False         False         False
38         False        False          True          True
39         False        False         False         False
40         False        False          True          True
41         False        False          True          True
42         False        False         False         False
43         False        False          True          True
44         False        False         False          True
45         False         True          True          True
46          True         True          True          True
47         False        False          True          True
48         False         True         False          True
49         False        False         False         False
50          True         True          True          True
51          True         True          True          True
52         False         True          True          True
53          True         True          True          True
54          True         True          True          True
55          True         True          True          True
56          True         True          True          True
57         False        False          True          True
58          True         True          True          True
59          True         True          True          True
60          True         True          True          True
61          True         True          True          True
62         False        False         False          True
63          True         True          True          True
64         False        False         False         False
65         False        False         False         False
66          True         True          True          True
67         False        False          True          True
68 

### Feature Importance

Another alternative to reduce the number of features is "to score input features using a model and use a filter-based feature selection method. These are called Feature Importance methods" (BROWNLEE, 2021). The most use Feature Importance methods are Classification and Regression Trees (CART), Random Forest, Bagged Decision Trees, and Gradient Boosting.

We use the Random Forest algorithm as our feature importance method. Decision tree algorithms, such as Random Forest, "offer importance scores based on the reduction in the criterion used to select split points, like Gini or entropy" (BROWNLEE, 2021).

We use the same number of features as in the previous selection methods. Thus, we select the 190, 150, 70, and 30 most relevant features and evaluate the models again. We see that the 30 more important features represent almost 80% of all importance. With the 70 most important features we reach 90%, and with the 150 most important 99%. 

**We can see that that KNN was the best model and that it performed similarly for the four cases tested**. Besides, the models performed a little worse with these features than with the features selected using the RFE method.

In [161]:
# random forest for feature importance on a regression problem
from sklearn.ensemble import RandomForestRegressor

# feature selection
def select_features_FI(X_train, y_train,):
    # configure to select all features
    RFR = RandomForestRegressor()
    # learn relationship from training data
    RFR.fit(X_train, y_train)
    # transform train input data
    importance = RFR.feature_importances_
    return importance

In [162]:
# feature selection
importance = select_features_FI(blog_X_train, blog_y_train)

# what are scores for the features
FI = pd.DataFrame(importance, columns = ['Importance'])
FI.sort_values(by=['Importance'],ascending=False).cumsum().reset_index()

index  Importance
0       60    0.233836
1       51    0.355880
2        1    0.405083
3        4    0.449059
4       19    0.485280
5       61    0.519114
6       21    0.546438
7       11    0.573639
8       54    0.600359
9        6    0.626309
10      20    0.649278
11      10    0.666500
12      50    0.681423
13      53    0.694962
14       3    0.704801
15     225    0.712857
16      56    0.720805
17      59    0.728462
18      14    0.735269
19      58    0.741730
20      13    0.748010
21      15    0.754278
22      68    0.760228
23     265    0.766074
24      48    0.771583
25      55    0.776372
26     100    0.780783
27     267    0.784977
28     273    0.789084
29     247    0.793145
30       8    0.796971
31     193    0.800745
32       0    0.804475
33     268    0.808118
34      18    0.811662
35      52    0.815029
36     274    0.818337
37     271    0.821633
38     101    0.824873
39     140    0.827683
40      63    0.830485
41     240    0.833249
42      24    0.835945
43     153    0.838620
44      95    0.841184
45     138    0.843732
46     246    0.846276
47      66    0.848689
48      78    0.851098
49     170    0.853447
50     232    0.855792
51     231    0.858124
52     229    0.860443
53     150    0.862756
54       9    0.865005
55       5    0.867245
56     264    0.869481
57      16    0.871715
58     157    0.873937
59     201    0.876113
60      57    0.878269
61     269    0.880423
62     121    0.882568
63     220    0.884705
64     139    0.886803
65     209    0.888868
66     181    0.890890
67     113    0.892910
68      22    0.894921
69     270    0.896918
70     119    0.898914
71     262    0.900910
72     107    0.902881
73     190    0.904847
74     212    0.906802
75     263    0.908749
76     192    0.910685
77     158    0.912601
78     215    0.914493
79     266    0.916382
80     194    0.918255
81     276    0.920116
82     179    0.921965
83     183    0.923780
84      67    0.925588
85     162    0.927383
86     227    0.929140
87     135    0.930883
88     272    0.932487
89      88    0.934080
90      23    0.935649
91     188    0.937148
92     218    0.938620
93     182    0.940084
94     279    0.941514
95      76    0.942940
96      45    0.944330
97     207    0.945717
98     233    0.947039
99     114    0.948344
100    137    0.949628
101    180    0.950878
102    206    0.952105
103     28    0.953303
104    196    0.954492
105    124    0.955618
106    244    0.956717
107    275    0.957786
108     30    0.958854
109    169    0.959918
110     46    0.960936
111     41    0.961906
112    133    0.962859
113     85    0.963804
114     92    0.964738
115    161    0.965650
116    253    0.966550
117    278    0.967437
118     47    0.968319
119     43    0.969183
120    102    0.970045
121     31    0.970904
122     35    0.971758
123    245    0.972595
124     25    0.973419
125    115    0.974219
126    191    0.975014
127     33    0.975793
128     36    0.976569
129     26    0.977345
130    120    0.978074
131    103    0.978779
132    204    0.979479
133    142    0.980153
134    256    0.980823
135    251    0.981484
136     38    0.982143
137    117    0.982798
138     40    0.983452
139     84    0.984100
140    152    0.984742
141    260    0.985380
142    109    0.986006
143     87    0.986609
144    134    0.987197
145    257    0.987780
146    254    0.988355
147     71    0.988921
148    118    0.989454
149    214    0.989977
150    195    0.990492
151    185    0.990960
152     44    0.991415
153    141    0.991856
154    136    0.992293
155      2    0.992722
156    217    0.993149
157     29    0.993569
158    143    0.993987
159     17    0.994370
160    174    0.994705
161    250    0.995026
162    156    0.995339
163    248    0.995637
164    226    0.995932
165    173    0.996203
166     97    0.996430
167    224    0.996653
168    163    0.996871
169    127    0.997084
170    230    0.997290
171    144    0.997457
172    205    0.997613
173

In [164]:
matriz_fi = pd.DataFrame(columns=['features','model','MAE_mean','MAE_std','RMSE_mean','RMSE_std','R2_mean','R2_std'])

for k in [30,70,150,190]:

    best_features_fi = FI.transpose()
    best_features_fi.columns = blog_X_train.columns
    best_features_fi.sort_values('Importance',axis=1,ascending=False,inplace=True)
    best_features_fi.drop(best_features_fi.iloc[:,k:],axis=1,inplace=True)
    blog_X_train_fi = blog_X_train[best_features_fi.columns]
  
    matriz_fi = estimator_cross_val_fea(k,'Linear Regression',LinearRegression(),     estimator_transf,matriz_fi,rs,blog_X_train_fi,blog_y_train)
    matriz_fi = estimator_cross_val_fea(k,'Ridge Regression', Ridge(),                estimator_transf,matriz_fi,rs,blog_X_train_fi,blog_y_train)
    matriz_fi = estimator_cross_val_fea(k,'Lasso',            Lasso(),                estimator_transf,matriz_fi,rs,blog_X_train_fi,blog_y_train)
    matriz_fi = estimator_cross_val_fea(k,'Elastic Net',      ElasticNet(),           estimator_transf,matriz_fi,rs,blog_X_train_fi,blog_y_train)
    matriz_fi = estimator_cross_val_fea(k,'KNN',              KNeighborsRegressor(),  estimator_transf,matriz_fi,rs,blog_X_train_fi,blog_y_train)
    matriz_fi = estimator_cross_val_fea(k,'CART',             DecisionTreeRegressor(),estimator_transf,matriz_fi,rs,blog_X_train_fi,blog_y_train)

matriz_fi

features              model  MAE_mean   MAE_std  RMSE_mean  RMSE_std  \
0        30  Linear Regression  9.215162  0.085426  30.288801  2.135071   
1        30   Ridge Regression  9.215138  0.085427  30.288797  2.135071   
2        30              Lasso  9.129967  0.102219  30.268826  2.153477   
3        30        Elastic Net  9.151172  0.101606  30.271307  2.153273   
4        30                KNN  6.390071  0.175214  28.600131  1.181100   
5        30               CART  6.346084  0.313916  33.064939  3.252297   
6        70  Linear Regression  9.319528  0.092706  30.332038  2.114239   
7        70   Ridge Regression  9.311305  0.087722  30.330393  2.114602   
8        70              Lasso  9.118657  0.099454  30.268890  2.161280   
9        70        Elastic Net  9.137884  0.096448  30.289800  2.149262   
10       70                KNN  6.404354  0.188587  28.632275  1.106485   
11       70               CART  6.504972  0.433481  34.278965  3.701548   
12      150  Linear Regression  9.448329  0.079941  30.371398  2.125482   
13      150   Ridge Regression  9.442058  0.076915  30.368486  2.124648   
14      150              Lasso  9.133528  0.098377  30.271053  2.163223   
15      150        Elastic Net  9.149090  0.095997  30.289858  2.148739   
16      150                KNN  6.390369  0.188168  28.570243  1.107732   
17      150               CART  6.463448  0.264339  34.543348  2.637511   
18      190  Linear Regression  9.505498  0.090742  30.381834  2.105945   
19      190   Ridge Regression  9.499521  0.088719  30.377565  2.105256   
20      190              Lasso  9.122078  0.093532  30.291712  2.144632   
21      190        Elastic Net  9.141841  0.092861  30.305065  2.136060   
22      190                KNN  6.390182  0.188812  28.577377  1.109084   
23      190               CART  6.426869  0.252181  34.310613  2.836365   

     R2_mean    R2_std  
0   0.352645  0.048617  
1   0.352645  0.048617  
2   0.353507  0.049235  
3   0.353381  0.049473  
4   0.421623  0.033184  
5   0.221103  0.144679  
6   0.350780  0.047987  
7   0.350843  0.048102  
8   0.353461  0.050071  
9   0.352580  0.049460  
10  0.419611  0.042727  
11  0.158598  0.182231  
12  0.349129  0.048015  
13  0.349244  0.048103  
14  0.353363  0.050215  
15  0.352572  0.049510  
16  0.422075  0.043243  
17  0.151365  0.126868  
18  0.348659  0.047507  
19  0.348836  0.047560  
20  0.352490  0.049360  
21  0.351926  0.048940  
22  0.421749  0.043835  
23  0.164378  0.121539

### Comparing the features

Finally, we compare the features selected by each model. We see that the percentage of features selected by all methods increase as the number of features used increases:

1. 30 best - 8 shared (26.7%)
2. 70 best - 29 shared (41.4%)
3. 150 best - 93 shared (62.0%)
4. 190 best - 135 shared (71.1%)

Besides, we can argue that the 8 features that were selected by all methods as one of the 30 most relevant fatures are the most significant ones for predicting our target variable.

In [185]:
MI_30 = MI.sort_values(by=['Score'],ascending=False)[:30].reset_index()
FI_30 = FI.sort_values(by=['Importance'],ascending=False)[:30].reset_index()
RF_30 = rfe_features.sort_values(by=['30 Features'],ascending=False)[:30].reset_index()
RF_30 = RF_30.drop(columns=['70 Features','150 Features','190 Features'])
merged_30 = pd.merge(FI_30, MI_30, on=['index'], how='inner')
merged_30 = pd.merge(merged_30, RF_30, on=['index'], how='inner')
merged_30

index  Importance     Score  30 Features
0     51    0.122044  0.275880         True
1      4    0.043976  0.178890         True
2     21    0.027324  0.209719         True
3     54    0.026720  0.237840         True
4     20    0.022969  0.202529         True
5     50    0.014923  0.174909         True
6     53    0.013539  0.170095         True
7      3    0.009840  0.186027         True
8     15    0.006268  0.194932         True

In [187]:
MI_70 = MI.sort_values(by=['Score'],ascending=False)[:70].reset_index()
FI_70 = FI.sort_values(by=['Importance'],ascending=False)[:70].reset_index()
RF_70 = rfe_features.sort_values(by=['70 Features'],ascending=False)[:70].reset_index()
RF_70 = RF_70.drop(columns=['30 Features','150 Features','190 Features'])
merged_70 = pd.merge(FI_70, MI_70, on=['index'], how='inner')
merged_70 = pd.merge(merged_70, RF_70, on=['index'], how='inner')
merged_70

index  Importance     Score  70 Features
0      60    0.233836  0.063339         True
1      51    0.122044  0.275880         True
2       1    0.049203  0.206831         True
3       4    0.043976  0.178890         True
4      61    0.033834  0.073060         True
5      21    0.027324  0.209719         True
6      54    0.026720  0.237840         True
7      20    0.022969  0.202529         True
8      50    0.014923  0.174909         True
9      53    0.013539  0.170095         True
10      3    0.009840  0.186027         True
11    225    0.008055  0.024111         True
12     56    0.007948  0.025834         True
13     59    0.007657  0.017393         True
14     58    0.006461  0.012028         True
15     15    0.006268  0.194932         True
16     68    0.005950  0.018598         True
17     48    0.005509  0.064550         True
18     55    0.004789  0.015041         True
19    100    0.004411  0.064659         True
20    247    0.004061  0.009429         True
21     52    0.003366  0.056459         True
22     24    0.002695  0.058171         True
23    138    0.002549  0.018655         True
24     66    0.002413  0.030007         True
25     78    0.002409  0.012075         True
26    231    0.002332  0.030719         True
27      9    0.002249  0.168995         True
28      5    0.002239  0.202373         True
29     22    0.002011  0.184062         True

In [189]:
MI_150 = MI.sort_values(by=['Score'],ascending=False)[:150].reset_index()
FI_150 = FI.sort_values(by=['Importance'],ascending=False)[:150].reset_index()
RF_150 = rfe_features.sort_values(by=['150 Features'],ascending=False)[:150].reset_index()
RF_150 = RF_150.drop(columns=['30 Features','70 Features','190 Features'])
merged_150 = pd.merge(FI_150, MI_150, on=['index'], how='inner')
merged_150 = pd.merge(merged_150, RF_150, on=['index'], how='inner')
merged_150

index  Importance     Score  150 Features
0      60    0.233836  0.063339          True
1      51    0.122044  0.275880          True
2       1    0.049203  0.206831          True
3       4    0.043976  0.178890          True
4      19    0.036221  0.173680          True
5      61    0.033834  0.073060          True
6      21    0.027324  0.209719          True
7      54    0.026720  0.237840          True
8       6    0.025950  0.207496          True
9      20    0.022969  0.202529          True
10     10    0.017222  0.203635          True
11     50    0.014923  0.174909          True
12     53    0.013539  0.170095          True
13      3    0.009840  0.186027          True
14    225    0.008055  0.024111          True
15     56    0.007948  0.025834          True
16     59    0.007657  0.017393          True
17     14    0.006806  0.135486          True
18     58    0.006461  0.012028          True
19     13    0.006280  0.188195          True
20     15    0.006268  0.194932          True
21     68    0.005950  0.018598          True
22    265    0.005846  0.003709          True
23     55    0.004789  0.015041          True
24    100    0.004411  0.064659          True
25    267    0.004194  0.002220          True
26    247    0.004061  0.009429          True
27      8    0.003826  0.190217          True
28    193    0.003775  0.006739          True
29      0    0.003730  0.207897          True
30     18    0.003545  0.183795          True
31     52    0.003366  0.056459          True
32    271    0.003296  0.002224          True
33     63    0.002802  0.007725          True
34     24    0.002695  0.058171          True
35     95    0.002564  0.005262          True
36    138    0.002549  0.018655          True
37    246    0.002543  0.005080          True
38     66    0.002413  0.030007          True
39     78    0.002409  0.012075          True
40    232    0.002346  0.008656          True
41    231    0.002332  0.030719          True
42    150    0.002313  0.008241          True
43      9    0.002249  0.168995          True
44      5    0.002239  0.202373          True
45    264    0.002237  0.001187          True
46     16    0.002234  0.200722          True
47    157    0.002222  0.029748          True
48    201    0.002176  0.006684          True
49    269    0.002154  0.002892          True
50    121    0.002145  0.007161          True
51    209    0.002064  0.017102          True
52    113    0.002020  0.010003          True
53     22    0.002011  0.184062          True
54    270    0.001997  0.003099          True
55    119    0.001996  0.006538          True
56    190    0.001967  0.005337          True
57    212    0.001954  0.039152          True
58    192    0.001936  0.002174          True
59    276    0.001862  0.001180          True
60    183    0.001815  0.003334          True
61     67    0.001807  0.007090          True
62    272    0.001604  0.000894          True
63     23    0.001569  0.191542          True
64    218    0.001472  0.002179          True
65    182    0.001464  0.004222          True
66     76    0.001425  0.004020          True
67     45    0.001391  0.076726          True
68    233    0.001322  0.004177          True
69    137    0.001284  0.002078          True
70    206    0.001226  0.002739          True
71    196    0.001189  0.001731          True
72     30    0.001068  0.107790          True
73    169    0.001064  0.001286          True
74     46    0.001017  0.114905          True
75     41    0.000970  0.108989          True
76    133    0.000953  0.001990          True
77     85    0.000945  0.001555          True
78     92    0.000934  0.001695          True
79     47    0.000882  0.050074          True
80     43    0.000864  0.058855          True
81    102    0.000862  0.003610          True
82     31    0.000859  0.119140          True
83     25    0.000824  0.115724          True
84    191    0.000794  0.001015          True
85     26    0.000776  0.118753          True
86

In [190]:
MI_190 = MI.sort_values(by=['Score'],ascending=False)[:190].reset_index()
FI_190 = FI.sort_values(by=['Importance'],ascending=False)[:190].reset_index()
RF_190 = rfe_features.sort_values(by=['190 Features'],ascending=False)[:190].reset_index()
RF_190 = RF_190.drop(columns=['30 Features','70 Features','150 Features'])
merged_190 = pd.merge(FI_190, MI_190, on=['index'], how='inner')
merged_190 = pd.merge(merged_190, RF_190, on=['index'], how='inner')
merged_190

index  Importance     Score  190 Features
0       60    0.233836  0.063339          True
1       51    0.122044  0.275880          True
2        1    0.049203  0.206831          True
3        4    0.043976  0.178890          True
4       19    0.036221  0.173680          True
5       61    0.033834  0.073060          True
6       21    0.027324  0.209719          True
7       11    0.027200  0.208136          True
8       54    0.026720  0.237840          True
9        6    0.025950  0.207496          True
10      20    0.022969  0.202529          True
11      10    0.017222  0.203635          True
12      50    0.014923  0.174909          True
13      53    0.013539  0.170095          True
14       3    0.009840  0.186027          True
15     225    0.008055  0.024111          True
16      56    0.007948  0.025834          True
17      59    0.007657  0.017393          True
18      14    0.006806  0.135486          True
19      58    0.006461  0.012028          True
20      13    0.006280  0.188195          True
21      15    0.006268  0.194932          True
22      68    0.005950  0.018598          True
23     265    0.005846  0.003709          True
24      48    0.005509  0.064550          True
25      55    0.004789  0.015041          True
26     100    0.004411  0.064659          True
27     267    0.004194  0.002220          True
28     273    0.004107  0.000501          True
29     247    0.004061  0.009429          True
30       8    0.003826  0.190217          True
31     193    0.003775  0.006739          True
32       0    0.003730  0.207897          True
33      18    0.003545  0.183795          True
34      52    0.003366  0.056459          True
35     271    0.003296  0.002224          True
36     140    0.002810  0.000591          True
37      63    0.002802  0.007725          True
38     240    0.002764  0.000478          True
39      24    0.002695  0.058171          True
40     153    0.002675  0.000229          True
41      95    0.002564  0.005262          True
42     138    0.002549  0.018655          True
43     246    0.002543  0.005080          True
44      66    0.002413  0.030007          True
45      78    0.002409  0.012075          True
46     232    0.002346  0.008656          True
47     231    0.002332  0.030719          True
48     150    0.002313  0.008241          True
49       9    0.002249  0.168995          True
50       5    0.002239  0.202373          True
51     264    0.002237  0.001187          True
52      16    0.002234  0.200722          True
53     157    0.002222  0.029748          True
54     201    0.002176  0.006684          True
55     269    0.002154  0.002892          True
56     121    0.002145  0.007161          True
57     139    0.002098  0.000240          True
58     209    0.002064  0.017102          True
59     113    0.002020  0.010003          True
60      22    0.002011  0.184062          True
61     270    0.001997  0.003099          True
62     119    0.001996  0.006538          True
63     107    0.001971  0.000606          True
64     190    0.001967  0.005337          True
65     212    0.001954  0.039152          True
66     192    0.001936  0.002174          True
67     215    0.001892  0.000021          True
68     276    0.001862  0.001180          True
69     179    0.001849  0.000266          True
70     183    0.001815  0.003334          True
71      67    0.001807  0.007090          True
72     272    0.001604  0.000894          True
73      23    0.001569  0.191542          True
74     218    0.001472  0.002179          True
75     182    0.001464  0.004222          True
76      76    0.001425  0.004020          True
77      45    0.001391  0.076726          True
78     233    0.001322  0.004177          True
79     137    0.001284  0.002078          True
80     206    0.001226  0.002739          True
81      28    0.001198  0.058288          True
82     196    0.001189  0.001731          True
83     244    0.001099  0.000000          True
84      30

----------

### Dimensionality reduction

"Dimensionality reduction refers to techniques that reduce the number of input variables in a dataset. Fewer input dimensions often mean correspondingly fewer parameters or a simpler structure in the machine learning model, referred to as degrees of freedom. A model with too many degrees of freedom is likely to overfit the training dataset and therefore may not perform well on new data" (BROWNLEE, 2021).

There are several techniques to reduce a data set dimensionality. **In this notebook, we use the Principal Component Analysis (PCA), which is the most used method for dimensionality reduction**. "It can be thought of as a projection method where data with m-columns (features) is projected into a subspace with m or fewer columns, whilst retaining the essence of the original data" (BROWNLEE, 2021). 

We reduce the dimensionality of the data set using the same number of features used in the feature selection section, 30, 70, 150, and 190. **We verify that the results are quite similar to the ones we obtained in the feature selection and that the level of the dimensionality reduction have little influence in the results**. Besides, once more the KNN model is better than the other models tested.

In [200]:
from sklearn.decomposition import PCA

def estimator_pca(estimator,k):
    #imputer = SimpleImputer(strategy='median')
    pipeline = Pipeline(steps=[('pca',PCA(n_components=k)),('model', estimator)])
    return pipeline 

def estimator_cross_val_pca(k,model,estimator,pipe,matriz,rs,X,y):
    pipe_ = pipe(estimator,k)
    scoring = ['neg_mean_absolute_error', 'neg_root_mean_squared_error','r2']
    kfold = KFold(n_splits=5, random_state=rs,shuffle=True)
    scores = cross_validate(pipe_,X,y,cv=kfold,scoring=scoring)
    
    mae_scores = -scores.get('test_neg_mean_absolute_error')
    mae_mean = mae_scores.mean()
    mae_std = mae_scores.std()
    
    rmse_scores = -scores.get('test_neg_root_mean_squared_error')
    rmse_mean = rmse_scores.mean()
    rmse_std = rmse_scores.std()
    
    r2_scores = scores.get('test_r2')
    r2_mean = r2_scores.mean()
    r2_std = r2_scores.std()
    
    results_ = [k,model,mae_mean,mae_std,rmse_mean,rmse_std,r2_mean,r2_std]
    results_ = pd.Series(results_, index = matriz.columns)
    results = matriz.append(results_,ignore_index=True)
    return results

matriz_pca = pd.DataFrame(columns=['dimensionality','model','MAE_mean','MAE_std','RMSE_mean','RMSE_std','R2_mean','R2_std'])

for k in [30,70,150,190]:
  
    matriz_pca = estimator_cross_val_pca(k,'Linear Regression',LinearRegression(),     estimator_pca,matriz_pca,rs,blog_X_train,blog_y_train)
    matriz_pca = estimator_cross_val_pca(k,'Ridge Regression', Ridge(),                estimator_pca,matriz_pca,rs,blog_X_train,blog_y_train)
    matriz_pca = estimator_cross_val_pca(k,'Lasso',            Lasso(),                estimator_pca,matriz_pca,rs,blog_X_train,blog_y_train)
    matriz_pca = estimator_cross_val_pca(k,'Elastic Net',      ElasticNet(),           estimator_pca,matriz_pca,rs,blog_X_train,blog_y_train)
    matriz_pca = estimator_cross_val_pca(k,'KNN',              KNeighborsRegressor(),  estimator_pca,matriz_pca,rs,blog_X_train,blog_y_train)
    matriz_pca = estimator_cross_val_pca(k,'CART',             DecisionTreeRegressor(),estimator_pca,matriz_pca,rs,blog_X_train,blog_y_train)

matriz_pca

dimensionality              model  MAE_mean   MAE_std  RMSE_mean  RMSE_std  \
0              30  Linear Regression  9.181340  0.086859  30.331096  2.126018   
1              30   Ridge Regression  9.181343  0.086857  30.331102  2.126015   
2              30              Lasso  9.129022  0.092453  30.294741  2.140151   
3              30        Elastic Net  9.152400  0.092614  30.305149  2.131907   
4              30                KNN  6.387640  0.183435  28.574896  1.109788   
5              30               CART  6.836427  0.497267  36.743567  3.679492   
6              70  Linear Regression  9.301259  0.082636  30.335895  2.124918   
7              70   Ridge Regression  9.299170  0.081941  30.336846  2.121471   
8              70              Lasso  9.129022  0.092453  30.294741  2.140151   
9              70        Elastic Net  9.151207  0.092924  30.305377  2.132085   
10             70                KNN  6.389464  0.189621  28.576422  1.107976   
11             70               CART  6.917342  0.304737  36.159541  2.953660   
12            150  Linear Regression  9.462956  0.090607  30.362443  2.110180   
13            150   Ridge Regression  9.461919  0.089732  30.362698  2.110798   
14            150              Lasso  9.129022  0.092453  30.294741  2.140151   
15            150        Elastic Net  9.151207  0.092924  30.305377  2.132085   
16            150                KNN  6.387296  0.189732  28.573651  1.106584   
17            150               CART  7.317104  0.255888  38.010531  3.331500   
18            190  Linear Regression  9.509379  0.089502  30.372373  2.107230   
19            190   Ridge Regression  9.508233  0.089958  30.372129  2.106809   
20            190              Lasso  9.129022  0.092453  30.294741  2.140151   
21            190        Elastic Net  9.151207  0.092924  30.305377  2.132085   
22            190                KNN  6.388800  0.190134  28.574231  1.108647   
23            190               CART  7.131948  0.302353  37.091357  3.460719   

     R2_mean    R2_std  
0   0.350834  0.048291  
1   0.350834  0.048291  
2   0.352380  0.048943  
3   0.351940  0.048552  
4   0.421874  0.043481  
5   0.042285  0.157737  
6   0.350627  0.048267  
7   0.350581  0.048194  
8   0.352380  0.048943  
9   0.351930  0.048574  
10  0.421791  0.043774  
11  0.070975  0.139547  
12  0.349499  0.047557  
13  0.349487  0.047592  
14  0.352380  0.048943  
15  0.351930  0.048574  
16  0.421894  0.043846  
17 -0.020517  0.120541  
18  0.349070  0.047495  
19  0.349079  0.047486  
20  0.352380  0.048943  
21  0.351930  0.048574  
22  0.421881  0.043761  
23  0.026693  0.135815

---------

# 3. Evaluate the ML model

Now evaluate the performance of our ML model in the test set, to see how it perform with unseen data.

We will do two tests. In the first one we use the KNN model and the 30 features selected using the RFE method. For the second test, we use the KNN model and the data set reduced using the PCA method.

First, we import the test set.

After testing the models, we verify that the performance of our model with the test set is similar to the performance with the train set. The MAE and RMSE are actually a little better but the R² is lower. Besides, the RMSE is considrably higher than the MAE. This result suggests that our data has many outliers and, consequently, our model is making some big errors.

Finally, we see that using the features selected by the RFE method and doing a dimensionality reduction using the PCA have similiar results.

### Getting the test set

In [230]:
import os
import glob

os.chdir(r"/Users/leuzinger/Dropbox/Data Science/Awari/Regressions/BlogFeedback/Test/")
filenames = [i for i in glob.glob("*.csv")]
df = [pd.read_csv(file, sep = ",", header=None,) 
      for file in filenames]

In [231]:
blog_test = df[0]

for i in range(1,len(df)):
    blog_test = blog_test.append(df[i]) 

blog_test.reset_index(drop=True,inplace=True)
blog_test.set_axis(att,axis=1,inplace=True)
blog_test.head()

blog_avg_total  blog_std_total  blog_min_total  blog_max_total  \
0        0.000000        0.000000             0.0             0.0   
1        0.000000        0.000000             0.0             0.0   
2        0.056075        0.330159             0.0             2.0   
3        0.000000        0.000000             0.0             0.0   
4       47.776787       93.737470             1.0           598.0   

   blog_median_total  blog_avg_last24h  blog_std_last24h  blog_min_last24h  \
0                0.0          0.000000          0.000000               0.0   
1                0.0          0.000000          0.000000               0.0   
2                0.0          0.018692          0.192442               0.0   
3                0.0          0.000000          0.000000               0.0   
4                7.5         17.857143         56.888218               0.0   

   blog_max_last24h  blog_median_last24h  blog_avg_24-48h  blog_std_24-48h  \
0               0.0                  0.0         0.000000         0.000000   
1               0.0                  0.0         0.000000         0.000000   
2               2.0                  0.0         0.018692         0.192442   
3               0.0                  0.0         0.000000         0.000000   
4             594.0                  1.0        17.350447        56.911470   

   blog_min_24-48h  blog_max_24-48h  blog_median_24-48h  blog_avg_first24h  \
0              0.0              0.0                 0.0           0.000000   
1              0.0              0.0                 0.0           0.000000   
2              0.0              2.0                 0.0           0.056075   
3              0.0              0.0                 0.0           0.000000   
4              0.0            594.0                 1.0          46.386160   

   blog_std_first24h  blog_min_first24h  blog_max_first24h  \
0           0.000000                0.0                0.0   
1           0.000000                0.0                0.0   
2           0.330159                0.0                2.0   
3           0.000000                0.0                0.0   
4          91.284140                1.0              595.0   

   blog_median_first24h  blog_avg_difference  blog_std_difference  \
0                   0.0             0.000000             0.000000   
1                   0.0             0.000000             0.000000   
2                   0.0             0.000000             0.273434   
3                   0.0             0.000000             0.000000   
4                   7.0             0.506696            79.062050   

   blog_min_difference  blog_max_difference  blog_median_difference  \
0                  0.0                  0.0                     0.0   
1                  0.0                  0.0                     0.0   
2                 -2.0                  2.0                     0.0   
3                  0.0                  0.0                     0.0   
4               -590.0                594.0                     0.0   

   blog_avg_total_tr  blog_std_total_tr  blog_min_total_tr  blog_max_total_tr  \
0           0.064516            0.24567                0.0                1.0   
1           0.000000            0.00000                0.0                0.0   
2           0.000000            0.00000                0.0                0.0   
3           0.064516            0.24567                0.0                1.0   
4           0.000000            0.00000                0.0                0.0   

   blog_median_total_tr  blog_avg_last24h_tr  blog_std_last24h_tr  \
0                   0.0             0.032258             0.176685   
1                   0.0             0.000000             0.000000   
2                   0.0             0.000000             0.000000   
3                   0.0             0.032258             0.176685   
4                   0.0             0.000000             0.000000   

   blog_min_last24h_tr  blog_max_last24h_tr  blog_median_last24h

In [232]:
blog_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7624 entries, 0 to 7623
Columns: 281 entries, blog_avg_total to target
dtypes: float64(281)
memory usage: 16.3 MB


In [233]:
blog_X_test = blog_test.drop('target',axis=1).copy()
blog_y_test = blog_test['target'].copy()

### Evaluating the ML models

In [234]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

rfe = RFE(estimator=DecisionTreeRegressor(), n_features_to_select=30) 
knn = KNeighborsRegressor()
pipe_rfe = Pipeline(steps=[('rfe',rfe),('knn',knn)])

pipe_rfe.fit(blog_X_train,blog_y_train)
blog_y_hat = pipe_rfe.predict(blog_X_test)

final_mae = mean_absolute_error(blog_y_test,blog_y_hat)
final_mse = mean_squared_error(blog_y_test,blog_y_hat)
final_rmse = np.sqrt(final_mse)
final_r2 = r2_score(blog_y_test,blog_y_hat)
print('MAE:  %.2f'%final_mae,'\nRMSE: %.2f'%final_rmse,'\nR2:   %.2f'%final_r2)

MAE:  5.79 
RMSE: 25.10 
R2:   0.32


In [235]:
pipe_pca = Pipeline(steps=[('pca',PCA(n_components=30)),('knn', KNeighborsRegressor())])
pipe_pca.fit(blog_X_train,blog_y_train)
blog_y_hat = pipe_pca.predict(blog_X_test)

final_mae = mean_absolute_error(blog_y_test,blog_y_hat)
final_mse = mean_squared_error(blog_y_test,blog_y_hat)
final_rmse = np.sqrt(final_mse)
final_r2 = r2_score(blog_y_test,blog_y_hat)
print('MAE:  %.2f'%final_mae,'\nRMSE: %.2f'%final_rmse,'\nR2:   %.2f'%final_r2)

MAE:  5.72 
RMSE: 25.10 
R2:   0.32


----------------------

## 4. Conclusion

In this notebook, we created a model to predict the number of blog posts in the next 24h based on several attributes of the post. First, we tested some regression models: 

1. Linear regression
2. Ridge regression
3. Lasso regression
4. Elastic Net
5. Classification and Regression Trees (CART)
6. k-Nearest Neighbors (KNN)

In this first tests, the KNN was the best performing method.

However, we verified that the large number of features in our data was demanding a high computing time to run the models. Therefore, we tested some techniques to reduce the number of features:

1. Mutual Information Statistics
2. Recursive Feature Elimination (RFE)
3. Random Forest

The features selected by the RFE were the ones that resulted in the best performance of the KNN model.

Finaly, we also used a dimensionality reduction method, the Principal Component Analysis (PCA) to reduce the size of our data set. Our results with the train set showed that both the RFE and the PCA, combined with the KNN model, had similar results.

**Therefore, we tested two models with our test set: (i) KNN + RFE and (ii) KNN + PCA. We verified that the models performed our identically**.

**However, our models performed modestly at best. All evaluation metrics used are poor**, especially the RMSE and the R². The fact that the RMSE is much higher than the suggests that our data has many outliers and, consequently, our model is making some big errors. **Nonetheless, given that these are quite simple regression methods, we could consider that the results are reasonable**. More complex models could be used to achieve better predictions. However, these models would probably demand more time to build and more computing power to run, which could actually mean a worse cost-benefit.